src/main/java/uk/gov/pay/api/model/CreatePaymentRequestBuilder.java

```diff
@@ -0,0 +1,181 @@
+package uk.gov.pay.api.model;
+
+import uk.gov.pay.commons.model.SupportedLanguage;
+import uk.gov.pay.commons.model.charge.ExternalMetadata;
+
+public class CreatePaymentRequestBuilder {
+    private ExternalMetadata metadata;
+    private int amount;
+    private String returnUrl;
+    private String reference;
+    private String description;
+    private String agreementId;
+    private SupportedLanguage language;
+    private Boolean delayedCapture;
+    private String email;
+    private String cardholderName;
+    private String addressLine1;
+    private String addressLine2;
+    private String city;
+    private String postcode;
+    private String country;
+    private PrefilledCardholderDetails prefilledCardholderDetails;
+
+    public CreatePaymentRequest build() {
+        return this.getAgreementId() == null ? new CreateCardPaymentRequest(this) : new CreateDirectDebitPaymentRequest(this);
```

Comment:
It feels weird to pass the builder to these constructors. It would be more conventional to have the builder extract the relevant fields from itself and then pass them to a multi-argument constructor. Alternatively, the builder could be renamed to something else (I suppose it’s a data source of some sort).

Reply:
I'm not keen on deconstructing the builder to pass to the constructors as it'll have too many arguments and be confusing - we'd also need a few overloaded constructors since some attributes are optional and then it just becomes a mess. I feel that passing a builder to a constructor is good practice and the builder _is_ a builder so I'm not entirely sure about renaming it.